modules/phase_field/src/materials/ComputeGBMisorientationType.C

```diff
@@ -0,0 +1,153 @@
+  params.addRequiredParam<UserObjectName>("ebsd_reader", "The EBSDReader GeneralUserObject");
+  params.addRequiredCoupledVarWithAutoBuild(
+      "v", "var_name_base", "op_num", "Array of coupled variables");
+  params.addRequiredParam<FileName>("file_name", "Misorientation angle data file name");
+  params.addParam<Real>("angle_threshold", 15, "Max LAGB Misorientation angle");
+  return params;
+}
+
+ComputeGBMisorientationType::ComputeGBMisorientationType(const InputParameters & parameters)
+  : Material(parameters),
+    _grain_tracker(getUserObject<GrainTracker>("grain_tracker")),
+    _ebsd_reader(getUserObject<EBSDReader>("ebsd_reader")),
+    _file_name(getParam<FileName>("file_name")),
+    _op_num(coupledComponents("v")),
+    _vals(coupledValues("v")),
+    _angle_threshold(getParam<Real>("angle_threshold")),
+    _gb_type(declareADProperty<Real>("gb_type"))
+{
+  readFile();
+}
+
+void
+ComputeGBMisorientationType::computeQpProperties()
+{
+
+  // Find out the number of boundary unique_id and save them
+  std::vector<unsigned int> gb_pairs;
+  std::vector<Real> gb_op_pairs;
+  MooseIndex(_vals) val_index = -1;
+
+  const auto & op_to_grains = _grain_tracker.getVarToFeatureVector(_current_elem->id());
+  for (auto & unique_id : op_to_grains)
+  {
+    val_index += 1;
+    if (unique_id == FeatureFloodCount::invalid_id)
+      continue;
+
+    auto grain_id = _ebsd_reader.getFeatureID(unique_id);
+
+    gb_pairs.push_back(grain_id);
+    gb_op_pairs.push_back((*_vals[val_index])[_qp]);
+  }
+
+  // Compute GB type by the number of id
+  switch (gb_pairs.size())
+  {
+    case 0:
+    {
+      break;
+    }
+    case 1:
+    {
+      _gb_type[_qp] = 0;
+      break;
+    }
+    case 2:
+    {
+      // get type by Misorientation angle
+      _gb_type[_qp] =
+          ((_misorientation_angles[getLineNum(gb_pairs[0], gb_pairs[1])] < _angle_threshold) ? 1
+                                                                                             : 2);
+      break;
+    }
+    default:
+      // get continuous type at triple junction
+      _gb_type[_qp] = getTripleJunctionType(gb_pairs, gb_op_pairs);
+  }
+}
+
+// Read the Misorientation angle file
+void
+ComputeGBMisorientationType::readFile()
+{
+  // Read in Euler angles from _file_name
+  std::ifstream inFile(_file_name.c_str());
+  if (!inFile)
+    mooseError("Can't open ", _file_name);
+
+  // Loop over misorientation angles
+  Real a;
+  while (inFile >> a)
+    _misorientation_angles.push_back(a);
+}
+
+// Function to output total line number of Misorientation angle file
+unsigned int
+ComputeGBMisorientationType::getTotalLineNum() const
+{
+  return _misorientation_angles.size();
+}
+
+// Function to output specific line number in Misorientation angle file
+unsigned int
+ComputeGBMisorientationType::getLineNum(unsigned int grain_i, unsigned int grain_j)
+{
+  if (grain_i > grain_j)
+    return grain_j + (grain_i - 1) * grain_i / 2;
```

Comment:
This does not look right.

Reply:
I suggest one of two things:

- drop this entirely. The misorientation angles of all pairs of grains are not fully independent parameters. You can calculate the misorientations from the individual grain orientations.
- or, (if you want to stick with this potentially inconsistent formulation), use a three column csv file with grain_i,grain_j,angle and put the angles into a map indexed by grain ID pairs.